File: GolangSolutions/Array/Q167_两数之和II输入有序数组.go
package Array

func twoSum(numbers []int, target int) []int {
	n := len(numbers)
	l, r := 0, n-1
	for l < r {
		sum := numbers[l] + numbers[r]
		if sum == target {
			return []int{l + 1, r + 1}
		} else if sum > target {
			r--
		} else {
			l++
		}
	}

	return []int{-1, -1}
}

func twoSum2(numbers []int, target int) []int {
	n := len(numbers)
	for i := range numbers {
		l, r := i+1, n-1
		for l <= r {
			m := l + (r-l)/2
			curTarget := target - numbers[i]
			if numbers[m] == curTarget {
				return []int{i + 1, m + 1}
			} else if numbers[m] > curTarget {
				r = m - 1
			} else {
				l = m + 1
			}
		}
	}

	return []int{-1, -1}
}

File: CPPSolutions/LeetCode/Array/Q167_两数之和II输入有序数组.cpp
#include <bits\stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int n = numbers.size();
        int low = 0, high = n - 1;
        while (low < high) {
            int sum = numbers[low] + numbers[high];
            if (sum == target) return vector<int>{ low + 1, high + 1 };
            else if (sum > target) high--;
            else low++;
        }

        return { -1, -1 };
    }

    vector<int> twoSum2(vector<int>& numbers, int target) {
        int n = numbers.size();
        for (int i = 0; i < n; i++) {
            int low = i + 1, high = n - 1;
            while (low <= high) {
                int mid = low + (high - low) / 2;
                int curTarget = target - numbers[i];
                if (numbers[mid] == curTarget) return { i + 1, mid + 1 };
                else if (numbers[mid] > curTarget) high = mid - 1;
                else low = mid + 1;
            }
        }

        return { -1, -1 };
    }
};